File: 22_专题/区间问题/UnionRectangle-包含原点的矩形面积并.py
"""
!`包含原点的`矩形合并/矩形面积并
维护一些矩形的面积并,支持查询面积和添加矩形。
注意矩形的范围在第一象限，且矩形左下角坐标为(0,0),右上角坐标为(x,y)。

三维情形:
扫描z轴,维护一个二维的矩形面积并。
https://kmjp.hatenablog.jp/entry/2017/06/20/0900
https://ei1333.github.io/library/structure/others/union-rectangle.hpp
"""

from typing import List
from sortedcontainers import SortedList

INF = int(1e18)


class UnionRectangleRange:
    __slots__ = ("_ur", "_ul", "_dr", "_dl")

    def __init__(self) -> None:
        self._ur = UnionRectangle()
        self._ul = UnionRectangle()
        self._dr = UnionRectangle()
        self._dl = UnionRectangle()

    def add(self, x1: int, x2: int, y1: int, y2: int) -> None:
        """Add [x1, x2] * [y1, y2].
        x1<=0<=x2.
        y1<=0<=y2.
        """
        assert x1 <= 0 <= x2, y1 <= 0 <= y2
        self._ur.add(x2, y2)
        self._ul.add(-x1, y2)
        self._dr.add(x2, -y1)
        self._dl.add(-x1, -y1)

    def query(self) -> int:
        """Return the sum of all rectangles."""
        return self._ur.sum + self._ul.sum + self._dr.sum + self._dl.sum


class UnionRectangle:
    __slots__ = ("_sl", "sum")

    def __init__(self) -> None:
        self._sl = SortedList([(0, INF), (INF, 0)])
        self.sum = 0

    def add(self, x: int, y: int) -> None:
        """Add [0, x] * [0, y]."""
        pos = self._sl.bisect_left((x, -INF))
        item = self._sl[pos]
        if item[1] >= y:
            return
        nextY = item[1]
        pos -= 1
        pre = self._sl[pos]
        while pre[1] <= y:
            x1, y1 = pre
            del self._sl[pos]
            pos -= 1
            pre = self._sl[pos]
            self.sum -= (x1 - pre[0]) * (y1 - nextY)
        self.sum += (x - self._sl[pos][0]) * (y - nextY)
        pos = self._sl.bisect_left((x, -INF))
        if self._sl[pos][0] == x:
            self._sl.pop(pos)
        self._sl.add((x, y))

    def query(self) -> int:
        """Return the sum of all rectangles."""
        return self.sum


if __name__ == "__main__":
    # points = [(2, 2), (3, 4), (1, 7)]
    # ur = UnionRectangleRange()
    # for x, y in points:
    #     ur.add(1, 3, 1, 3)
    #     print(ur.query())

    # https://yukicoder.me/problems/2577
    n = int(input())
    adds = [tuple(map(int, input().split())) for _ in range(n)]  # (x1,y1,x2,y2)
    UR = UnionRectangleRange()
    pre = 0
    for x1, y1, x2, y2 in adds:
        UR.add(x1, x2, y1, y2)
        cur = UR.query()
        print(cur - pre)
        pre = cur

    class Solution:
        def rectangleArea(self, rectangles: List[List[int]]) -> int:
            ur = UnionRectangleRange()
            for x1, y1, x2, y2 in rectangles:
                ur.add(x1, x2, y1, y2)
            return ur.query() % int(1e9 + 7)

File: 22_专题/区间问题/UnionRectangle-包含原点的矩形面积并.ts
/* eslint-disable @typescript-eslint/no-non-null-assertion */

import { SortedList } from '../离线查询/根号分治/SortedList/_SortedList'

const INF = 2e15

/**
 * !`包含原点的`矩形合并/矩形面积并.
 */
class UnionRectangleRange {
  private readonly _ur = new UnionRectangle()
  private readonly _ul = new UnionRectangle()
  private readonly _dr = new UnionRectangle()
  private readonly _dl = new UnionRectangle()

  /**
   * Add [x1, x2] * [y1, y2].
   * x1 <= 0 <= x2.
   * y1 <= 0 <= y2.
   */
  add(x1: number, x2: number, y1: number, y2: number): void {
    this._ur.add(x2, y2)
    this._ul.add(-x1, y2)
    this._dr.add(x2, -y1)
    this._dl.add(-x1, -y1)
  }

  query(): number {
    return this._ur.sum + this._ul.sum + this._dr.sum + this._dl.sum
  }
}

/**
 * !`包含原点的`矩形合并/矩形面积并.
 */
class UnionRectangle {
  sum = 0
  private readonly _sl = new SortedList(
    [
      [0, INF],
      [INF, 0]
    ],
    (a, b) => a[0] - b[0]
  )

  /**
   * Add [0, x] * [0, y].
   * x >= 0.
   * y >= 0.
   */
  add(x: number, y: number): void {
    let pos = this._sl.bisectLeft([x, -INF])
    const item = this._sl.at(pos)!
    if (item[1] >= y) return
    const nextY = item[1]
    pos--
    let pre = this._sl.at(pos)!
    while (pre[1] <= y) {
      const x1 = pre[0]
      const y1 = pre[1]
      this._sl.pop(pos)
      pos--
      pre = this._sl.at(pos)!
      this.sum -= (x1 - pre[0]) * (y1 - nextY)
    }
    this.sum += (x - this._sl.at(pos)![0]) * (y - nextY)
    pos = this._sl.bisectLeft([x, -INF])
    if (this._sl.at(pos)![0] === x) this._sl.pop(pos)
    this._sl.add([x, y])
  }

  query(): number {
    return this.sum
  }
}

if (require.main === module) {
  const R = new UnionRectangleRange()
  R.add(-2, 1, -2, 1)
  console.log(R.query())

  // https://yukicoder.me/submissions/857898
  // import * as fs from 'fs'
  // import { resolve } from 'path'

  // function useInput(path?: string) {
  //   let data: string
  //   if (path) {
  //     data = fs.readFileSync(resolve(__dirname, path), 'utf8')
  //   } else {
  //     data = fs.readFileSync(process.stdin.fd, 'utf8')
  //   }

  //   const lines = data.split(/\r\n|\r|\n/)
  //   let lineId = 0
  //   const input = (): string => lines[lineId++]

  //   return {
  //     input
  //   }
  // }

  // const { input } = useInput()
  // const n = Number(input())
  // const R = new UnionRectangleRange()
  // let pre = 0
  // for (let i = 0; i < n; i++) {
  //   const [x1, x2, y1, y2] = input().split(' ').map(Number)
  //   R.add(x1, x2, y1, y2)
  //   const cur = R.query()
  //   console.log(cur - pre)
  //   pre = cur
  // }
}

export { UnionRectangleRange, UnionRectangle }
